corehq/apps/app_manager/app_translations/generators.py

```diff
@@ -190,37 +189,25 @@ def _metadata(self):
             'Version': self.version
         }
 
-    def generate_translation_files(self):
-        self._build_translations()
-        self.po_file_generator = PoFileGenerator(self.translations, self._metadata)
 
-    @property
-    def generated_files(self):
-        if self.po_file_generator:
-            return self.po_file_generator.generated_files
-        return []
+class PoFileGenerator(object):
+    def __init__(self, translations, metadata):
+        self.translations = translations
+        self.metadata = metadata
 
-    def cleanup(self):
-        if self.po_file_generator:
-            self.po_file_generator.cleanup()
+    def __enter__(self):
+        return self
 
+    def __exit__(self, *exc_info):
+        self._cleanup()
 
-class PoFileGenerator(object):
-    def __init__(self, translations, metadata):
-        self.generated_files = list()  # list of tuples (filename, filepath)
```

Comment:
Removing this will break `self._cleanup()`. Recommendation: convert this class (`PoFileGenerator`) into a `contextmanager`-decorated function:
```py
from contextlib import contextmanager

@contextmanager
def generate_po_files(translations, metadata):
    """Generate PO files from translations

    Usage:

        with generate_translation_files(trans, meta) as files:
            for name, path in files:
                # do stuff with name, path
        # generated files are cleaned up (deleted) on with block exit

    :param translations: dict of translations:
    `{filename: [list of translations], ...}`
    :param metadata: PO file metadata.
    :yields: (context object) list of `POFileInfo` objects.
    """
    generated_files = []
    try:
        for file_name in translations:
            sheet_translations = translations[file_name]
            po = polib.POFile()
            po.check_for_duplicates = False
            po.metadata = metadata
            for translation in sheet_translations:
                source = translation.key
                if source:
                    entry = polib.POEntry(
                        msgid=translation.key,
                        msgstr=translation.translation or '',
                        occurrences=translation.occurrences,
                        msgctxt=translation.msgctxt
                    )
                    po.append(entry)
            temp_file = tempfile.NamedTemporaryFile(delete=False)
            po.save(temp_file.name)
            generated_files.append(POFileInfo(file_name, temp_file.name))
        yield generated_files
    finally:
        for resource_name, filepath in generated_files:
            if os.path.exists(filepath):
                os.remove(filepath)
```

Reply:
Ah right - that'd work.  You could also keep around a reference to the files as a private variable.

Reply:
Oh! sorry for the oversight. Got carried away with removing code. Adding as a private variable.

I would still like to keep this as a class.. There is quite a lot happening here to make it just a method and i would like to keep it grouped together, but there is definitely something to learn from that example. Thanks @millerdev